wrapper/src/main/java/software/amazon/jdbc/plugin/efm2/MonitorImpl.java

```diff
@@ -0,0 +1,418 @@
+  private final TelemetryCounter abortedConnectionsCounter;
+
+  /**
+   * Store the monitoring configuration for a connection.
+   *
+   * @param pluginService             A service for creating new connections.
+   * @param hostSpec                  The {@link HostSpec} of the server this {@link MonitorImpl}
+   *                                  instance is monitoring.
+   * @param properties                The {@link Properties} containing additional monitoring
+   *                                  configuration.
+   */
+  public MonitorImpl(
+      final @NonNull PluginService pluginService,
+      final @NonNull HostSpec hostSpec,
+      final @NonNull Properties properties,
+      final int failureDetectionTimeMillis,
+      final int failureDetectionIntervalMillis,
+      final int failureDetectionCount,
+      final TelemetryCounter abortedConnectionsCounter) {
+
+    this.pluginService = pluginService;
+    this.telemetryFactory = pluginService.getTelemetryFactory();
+    this.hostSpec = hostSpec;
+    this.properties = properties;
+    this.failureDetectionTimeNano = TimeUnit.MILLISECONDS.toNanos(failureDetectionTimeMillis);
+    this.failureDetectionIntervalNano = TimeUnit.MILLISECONDS.toNanos(failureDetectionIntervalMillis);
+    this.failureDetectionCount = failureDetectionCount;
+    this.abortedConnectionsCounter = abortedConnectionsCounter;
+
+    final String nodeId = StringUtils.isNullOrEmpty(this.hostSpec.getHostId())
+        ? this.hostSpec.getHost()
+        : this.hostSpec.getHostId();
+
+    this.newContextsSizeGauge = telemetryFactory.createGauge(
+        String.format("efm2.newContexts.size.%s", nodeId),
+        this::getActiveContextSize);
+
+    this.activeContextsSizeGauge = telemetryFactory.createGauge(
+        String.format("efm2.activeContexts.size.%s", nodeId),
+        () -> (long) this.activeContexts.size());
+
+    this.nodeHealtyGauge = telemetryFactory.createGauge(
+        String.format("efm2.nodeHealthy.%s", nodeId),
+        () -> this.nodeUnhealthy ? 0L : 1L);
+
+    this.threadPool.submit(this::newContextRun); // task to handle new contexts
+    this.threadPool.submit(this); // task to handle active monitoring contexts
+    this.threadPool.shutdown(); // No more tasks are accepted by pool.
+  }
+
+  @Override
+  public boolean canDispose() {
+    return this.activeContexts.size() == 0;
+  }
+
+  @Override
+  public void close() throws Exception {
+    this.stopped.set(true);
+
+    // Waiting for 30s gives a thread enough time to exit monitoring loop and close database connection.
+    if (!this.threadPool.awaitTermination(30, TimeUnit.SECONDS)) {
+      this.threadPool.shutdownNow();
+    }
+    LOGGER.finest(() -> Messages.get(
+        "MonitorImpl.stopped",
+        new Object[] {this.hostSpec.getHost()}));
+  }
+
+  protected long getActiveContextSize() {
+    return this.newContexts.values().stream()
+        .map(x -> (long) x.size())
+        .reduce(0L, Long::sum);
```

Comment:
You can also do this
```suggestion
    return this.newContexts.values().stream().mapToLong(Collection::size).sum();
```